modules/unspents/src/dimensions.ts

```diff
@@ -1,723 +1,531 @@
 
-  plus(v: Partial<IDimensions>): IDimensions;
-  times(n: number): IDimensions;
+/**
+ * Dimensions of a BitGo wallet transactions.
+ */
+export class Dimensions {
+  /** Input counts for BitGo wallet multi-signature inputs */
+  public readonly nP2shInputs: number = 0;
+  public readonly nP2shP2wshInputs: number = 0;
+  public readonly nP2wshInputs: number = 0;
+  public readonly nP2trKeypathInputs: number = 0;
+  public readonly nP2trScriptPathLevel1Inputs: number = 0;
+  public readonly nP2trScriptPathLevel2Inputs: number = 0;
 
-  isSegwit(): boolean;
-  getNInputs(): number;
-  getOverheadVSize(): number;
-  getInputsVSize(): number;
-  getOutputsVSize(): number;
-  getVSize(): number;
-}
+  /* Input count for single-signature inputs (Replay Protection inputs) */
+  public readonly nP2shP2pkInputs: number = 0;
 
-interface IFromInputParams {
-  // In cases where the input type is ambiguous, we must provide a hint about spend script type.
-  assumeUnsigned?: symbol;
-}
+  public readonly outputs: OutputDimensions = new OutputDimensions();
 
-export interface IFromUnspentParams {
-  // The p2tr output type has multiple spend options and thus different weights per spend path.
-  p2trSpendType: 'keypath' | 'scriptpath-level1' | 'scriptpath-level2';
-}
+  constructor(d: Partial<Dimensions> = {}) {
+    Object.entries(d).forEach(([key, value]) => this.setProperty(key, value));
 
-export interface IDimensionsStruct extends t.Struct<IDimensions> {
-  (v: IBaseDimensions): IDimensions;
-  ASSUME_P2SH: symbol;
-  ASSUME_P2SH_P2WSH: symbol;
-  ASSUME_P2WSH: symbol;
-  ASSUME_P2TR_KEYPATH: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL1: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL2: symbol;
-
-  SingleOutput: {
-    p2sh: IDimensions;
-    p2shP2wsh: IDimensions;
-    p2wsh: IDimensions;
-    p2tr: IDimensions;
-    p2pkh: IDimensions;
-    p2wpkh: IDimensions;
-  };
-  new (v: IBaseDimensions): IDimensions;
+    Object.freeze(this);
+  }
 
-  zero(): IDimensions;
-  sum(...args: Array<Partial<IDimensions>>): IDimensions;
+  private setProperty(k: string, v: unknown): void {
+    switch (k) {
+      case 'nP2shInputs':
+      case 'nP2shP2wshInputs':
+      case 'nP2wshInputs':
+      case 'nP2trKeypathInputs':
+      case 'nP2trScriptPathLevel1Inputs':
+      case 'nP2trScriptPathLevel2Inputs':
+      case 'nP2shP2pkInputs':
+        if (typeof v !== 'number') {
+          throw new Error(`property ${k} must be number`);
+        }
+        if (!Number.isSafeInteger(v) || v < 0) {
+          throw new Error(`property ${k} must be zero or positive integer`);
+        }
+        break;
+      case 'outputs':
+        if (!(v instanceof OutputDimensions)) {
+          v = new OutputDimensions(v as OutputDimensions);
+        }
+        break;
+      default:
+        throw new Error(`unknown property ${k}`);
+    }
 
-  getOutputScriptLengthForChain(chain: ChainCode): number;
-  getVSizeForOutputWithScriptLength(length: number): number;
+    (this as any)[k] = v;
```

Comment:
Ah interesting so this cast to any gets around the `readonly` type restriction internally without having to put `setProperty` in the constructor.